Feature/a2f/aspects on file leve

MLanghammer opened the pull request:
The PR contains the transformation implementation for autosar to franca on file level.
The major parts of the PR are:
-  the correct creation of franca files and their namespaces for interfaces and data types
- refactoring of tests
- refactoring of transformation logic in order to deal with the fact that we can create multiple franca files from a single autosar file

TorstenGoerg commented:
- "xtend-gen" should not be removed from the classpath of the plugin project "org.genivi.faracon.cli" as there a .xtend source file in this project.
- In case of class definitions with annotations, comments should not be placed between and annotations and the keyword "class", but in front of the annotations.
- Why should the show case example be excluded from the tests?
- ARA2FrancaTestBase: The Franca and AR factories are instantiated in 'Franca2ARATestBase' as well. This could be extracted to a common base class 'TestBase' of 'Franca2ARATestBase' and 'ARA2FrancaTestBase'. The same is true for the instantiation of 'FrancaPersistenceManager' and 'ARAConnector' which will be needed for the testing of both conversion directions as well.
- ARA2FrancaTestBase: The formal parameter 'path' should be more precisely named 'sourceFilePath' respectively 'expectedFilePath' to clarify the according meanings.
- ARA2FrancaTestBase: The method 'getTestPath()' could be extracted to the common base class 'TestBase' mentioned above.
- ARA2FrancaTransformation.transform(): AUTOSAR namespaces should not result in an error if they match the according hierarchy of packages.
- ConverterCliCommand.convertARAFiles(): The output directory preference should not be read in the loop for the Franca output model. If the output directory is not properly set multiple identical error messages are logged.
- Each test class should have a class comment to describe its purpose, i.e., as the class names are just based on the Franca Feature IDs.
- FrancaMultiModelContainer: The base class 'FrancaModelContainer' doesn't seem to be approriate for the composite design pattern. I.e., most of its methods deal with just one model.
- ARA2FrancaTestBase.transformAndCheck(String sourceFilePath, Collection<String> expectedFilePaths): Each expected Franca file is loaded into a separate resource set. This might cause problems in case of cross references between the models that result from the expected files.
- ARA2FrancaTestBase.transformToFranca(): Each Franca conversion output model is saved via a separate resource set. This might cause problems in case of cross references between the models.
- ARA2FrancaTestBase.saveAraFile(): Files created during test execution should not be written into the folder "src".
- IDL1100_Tests.testLeafPackageCreation(), IDL1130_Tests.testOneInterface(), and IDL1130_Tests. testMultipleInterfacesInOnePackage(): Why are the paths of the source and of the expected file handled differently? Both files are located in the same folder.
- Why is the class 'FrancaAraAssertHelper' separated from 'FrancaAssertHelper'? It is not ARA specific.
- IDL1110_Tests.testFrancaFileNameForSingleFileInput() and IDL1110_Tests.testFrancaFileNameForMultiFileOutput(): Files created during test execution should not be written into the main folder of the tests project.
- "namespacesHierarchy.fidl", "multiPackagesWithContent_parentPackage.firstPackageWithContent.fidl", and "multiPackagesWithContent_parentPackage.secondPackageWithContent.fidl": The main folder of the tests project is not a proper location for Franca model files.

MLanghammer commented:
Hi, answers inline,
> 
>     * In case of class definitions with annotations, comments should not be placed between and annotations and the keyword "class", but in front of the annotations.
I agree, however I was only found the class "ShowcaseReverseTransformationTests" where this is true? Could you point me to other classes, if any?

> 
>     * Why should the show case example be excluded from the tests?
It was on @Ignore already, i.e. it was not executed at all. I commented the test code out as it does not work with the new implementation of the Ara 2 Franca test enviroment.
So I guess we either need to delete the test class itself or try to make it run again. 
> 
>     * ARA2FrancaTestBase: The Franca and AR factories are instantiated in 'Franca2ARATestBase' as well. This could be extracted to a common base class 'TestBase' of 'Franca2ARATestBase' and 'ARA2FrancaTestBase'. The same is true for the instantiation of 'FrancaPersistenceManager' and 'ARAConnector' which will be needed for the testing of both conversion directions as well.
> 
*Yes, I created the base class now.

>     * ARA2FrancaTestBase: The formal parameter 'path' should be more precisely named 'sourceFilePath' respectively 'expectedFilePath' to clarify the according meanings.
> *Done.
>     * ARA2FrancaTestBase: The method 'getTestPath()' could be extracted to the common base class 'TestBase' mentioned above.
> Done
>     * ARA2FrancaTransformation.transform(): AUTOSAR namespaces should not result in an error if they match the according hierarchy of packages.
> 
> Even thouth this is a different in the spec I agree. Especially since we create the namespace from franca to autosar. I have done that now.
> 
>     * ConverterCliCommand.convertARAFiles(): The output directory preference should not be read in the loop for the Franca output model. If the output directory is not properly set multiple identical error messages are logged.
> 
> Done.
> 
>     * Each test class should have a class comment to describe its purpose, i.e., as the class names are just based on the Franca Feature IDs.
> 
* Done.
>
>     * FrancaMultiModelContainer: The base class 'FrancaModelContainer' doesn't seem to be approriate for the composite design pattern. I.e., most of its methods deal with just one model.
> 
> Thats correct, but I could not come up with a better solution to be able to deal with multiple models. To correct this, as far as I can see, we would need to change the Franca stuff. Do you have any other ideas? I added a comment to the class FrancaMultiModelContainer to explain the issue.
> 
>     * ARA2FrancaTestBase.transformAndCheck(String sourceFilePath, Collection expectedFilePaths): Each expected Franca file is loaded into a separate resource set. This might cause problems in case of cross references between the models that result from the expected files.
> 
>     * ARA2FrancaTestBase.transformToFranca(): Each Franca conversion output model is saved via a separate resource set. This might cause problems in case of cross references between the models.
> 
* Guess you are right. Was not an issue up until now. So my suggestion is to deal with it once it becomes relevant.

>     * ARA2FrancaTestBase.saveAraFile(): Files created during test execution should not be written into the folder "src".
> 
* I use this method to create the input ARXML files in a manual step and then store them into src. If you want we can change that. The fact that they are stored in src just comes in handy.

>     * IDL1100_Tests.testLeafPackageCreation(), IDL1130_Tests.testOneInterface(), and IDL1130_Tests. testMultipleInterfacesInOnePackage(): Why are the paths of the source and of the expected file handled differently? Both files are located in the same folder.
> 
* I guess I was trying to use the fidl files from f2 first, which turned out that this does not work in that case and then changed the path wrong...
> 
>     * Why is the class 'FrancaAraAssertHelper' separated from 'FrancaAssertHelper'? It is not ARA specific.
> 
* Right. My idea was to have a common assert helper for ara 2 franca and franca 2 ara and specific ones: one for franca and one for Ara. I renamed the FrancaAraAssertHelper to FaraconAssertHelper, which should make this more clear I hope. What do you think?
> 
>     * IDL1110_Tests.testFrancaFileNameForSingleFileInput() and IDL1110_Tests.testFrancaFileNameForMultiFileOutput(): Files created during test execution should not be written into the main folder of the tests project.
> 
>     * "namespacesHierarchy.fidl", "multiPackagesWithContent_parentPackage.firstPackageWithContent.fidl", and "multiPackagesWithContent_parentPackage.secondPackageWithContent.fidl": The main folder of the tests project is not a proper location for Franca model files.

* Right. The files are also not necessary. I committed them accidentialy.
